docs/configs-boot-tests/sample.py

```diff
@@ -0,0 +1,305 @@
+              (system.cpu). After this function the detailed CPU (timingCPU)
+              is active.
+
+        @param system the system we are running
+        @param warmup_time amount of guest-time to warmup
+        @param detailed_warmup_time amout of guest-time for detailed warming
+        @param sim_time amount of guest-time for detailed simulation
+    """
+    print "Warming up"
+    system.switchCpus(system.cpu, system.atomicCpu)
+    runSim(warmup_time)
+    print "Detailed warmup"
+    system.switchCpus(system.atomicCpu, system.timingCpu)
+    runSim(detailed_warmup_time)
+    print "Running the detailed simulation!"
+    # Reset the stats so we only record things in the detailed simulation
+    m5.stats.reset()
+    runSim(sim_time)
+
+def simulateROI(system, instructions, samples):
+    """ Do everything we need to do during the ROI.
+        At a high-level, this function executes the entire ROI and takes
+        a number of random sample points. At each point it may run multiple
+        simulations.
+
+        @param system the system we are running
+        @param instructions total instructions in the ROI (summed over all
+               CPUs)
+        @param samples number of sample points in the ROI
+    """
+    from random import randint, sample, seed
+    import signal
+    import time
+
+    seed()
+
+    executed_insts = system.totalInsts()
+
+    # We probably want to updated the max instructions by subtracting the
+    # insts that we are going to execute in the last detailed sim.
+    sample_secs = toLatency("10ms") + toLatency("10us") + toLatency("100us")
+    # assume 1 IPC per CPU
+    instructions -= int(sample_secs * system.clk_domain.clock[0].frequency *
+                        len(system.cpu) * 1)
+
+    # Get the instruction numbers that we want to exit at for our sample
+    # points. It needs to be sorted, too.
+    sample_insts = sample(xrange(executed_insts,
+                                 executed_insts + instructions),
+                          samples)
+    sample_insts.sort()
+
+    # These are the currently running processes we have forked.
+    pids = []
+
+    # Signal handler for when the processes exit. This is how we know when
+    # to remove the child from the list of pids.
+    def handler(signum, frame):
+        assert(signum == signal.SIGCHLD)
+        try:
+            while 1:
+                pid,status = os.wait()
+                if status != 0:
+                    print "pid", pid, "failed!"
+                    sys.exit(status)
+                pids.remove(pid)
+        except OSError:
+            pass
+
+    # install the signal handler
+    signal.signal(signal.SIGCHLD, handler)
+
+    # Here's the magic
+    for i, insts in enumerate(sample_insts):
+        print "Fast forwarding to sample", i, "stopping at", insts
+        insts_past = fastforward(system, insts - system.totalInsts())
+        if insts_past/insts > 0.01:
```

Comment:
Hi Jason, I wonder if the scheduleInstStop(0, exit_insts, "Max Insts") will result in the exceeded instrutions(real insts > exit_insts) when m5.simulate() stop?

Reply:
I'm not sure if I understand your question. However, I think the comments in the fastforward function are pretty clear: https://github.com/darchr/gem5art/blob/master/docs/gem5-configs/configs-boot-tests/sample.py#L61. When running with multiple CPUs you can only set the `scheduleInstStop` on a per-CPU basis, so it's imprecise if you want to stop after a certain number of global instructions.

In practice, I don't think I've ever seen `insts_past/insts > 0.01`.